src/nlsolve/utils.jl

```diff
@@ -0,0 +1,25 @@
+@inline function nlsolve(nlsolver::NLType, integrator::IntegType) where {NLType, IntegType}
+    maxiters = nlsolver.maxiters
+    dt = nlsolver.dt
+    p = integrator.p
+    t = nlsolver.t
+    γ = nlsolver.γ
+    tmp = nlsolver.tmp
+    z_i = nlsolver.z
+    c = nlsolver.c
+
+    abstol = 100eps(eltype(z_i))
+
+    for i in 1:maxiters
+        W_eval = nlsolver.W(tmp + γ * z_i, p, t + c * dt)
+        f_eval = integrator.f(tmp + γ * z_i, p, t + c * dt)
+        f_rhs = dt * f_eval - z_i
+        Δz = W_eval \ f_rhs
```

Comment:
no factorization reuse?

Reply:
I'd just implemented the simple Newton raphson here: we can leverage factorization reuse in quasi-newton schemes.
However, I saw factorizing and solving to be slow with Static Arrays on low dimensions: Even the Rodas schemes do not factorize within GPU kernels.

Reply:
Yeah but the main win of the Newton methods is the reuse of the factorization, which is the big difference from the Rosenbrock methods.

Reply:
Do you mean something like this?
One can use this type of quasi-scheme:

```
    W_eval = nlsolver.W(tmp + γ * z_i, p, t + c * dt)

    W_fact = lu(W_eval)

    new_W_iter = maxiters/3
    for i in 1:maxiters
        if iszero(i%new_W_iter)
            W_eval = nlsolver.W(tmp + γ * z_i, p, t + c * dt)
            W_fact = lu(W_eval)
        end
        f_eval = integrator.f(tmp + γ * z_i, p, t + c * dt)
        f_rhs = dt * f_eval - z_i
        Δz = W_fact \ f_rhs
        z_i = z_i - Δz

        if norm(dt * integrator.f(tmp + γ * z_i, p, t + c * dt) - z_i) < abstol
            break
        end
    end
    @set! nlsolver.z = z_i
end
```
Benchmark:

```
BenchmarkTools.Trial: 11 samples with 1 evaluation.
 Range (min … max):  486.748 ms … 513.772 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     487.229 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   490.556 ms ±   8.144 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                              
  ██▁▅▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅ ▁
  487 ms           Histogram: frequency by time          514 ms <

 Memory estimate: 6.45 KiB, allocs estimate: 146.

```

Before:

```
BenchmarkTools.Trial: 143 samples with 1 evaluation.
 Range (min … max):  31.773 ms … 83.528 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     32.301 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   35.021 ms ±  7.423 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▆                                                           
  ██▄▄▁▄▄▁▄▁▁▁▄▁▁▁▄▁▁▁▄▁▁▄▁▆▄▇▆▄▁▁▁▁▁▁▄▁▁▁▁▁▄▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ ▄
  31.8 ms      Histogram: log(frequency) by time      63.8 ms <

 Memory estimate: 6.48 KiB, allocs estimate: 147.

```

I am trying to avoid LU factorization since it's pretty expensive within a GPU kernel.

Reply:
WIll follow up with adding a size cutoff for LU factorization in the next PR.